File: mobile/lib/services/notification_service.dart
import 'package:dio/dio.dart';
import '../models/notification.dart';
import '../utils/constants.dart';
import 'api_service.dart';

class NotificationService {
  final ApiService _apiService = ApiService();

  // Get paginated notifications
  Future<NotificationPage> getNotifications({
    int page = 0,
    int size = 20,
  }) async {
    try {
      final response = await _apiService.get(
        ApiConstants.notifications,
        queryParameters: {
          'page': page,
          'size': size,
        },
      );

      return NotificationPage.fromJson(response.data);
    } catch (e) {
      throw _handleError(e);
    }
  }

  // Get unread notifications
  Future<List<AppNotification>> getUnreadNotifications() async {
    try {
      final response = await _apiService.get(
        '${ApiConstants.notifications}/unread',
      );

      return (response.data as List)
          .map((item) => AppNotification.fromJson(item))
          .toList();
    } catch (e) {
      throw _handleError(e);
    }
  }

  // Get unread count
  Future<int> getUnreadCount() async {
    try {
      final response = await _apiService.get(
        '${ApiConstants.notifications}/unread/count',
      );

      return response.data['count'] as int;
    } catch (e) {
      throw _handleError(e);
    }
  }

  // Mark notification as read
  Future<void> markAsRead(int notificationId) async {
    try {
      await _apiService.put(
        '${ApiConstants.notifications}/$notificationId/read',
      );
    } catch (e) {
      throw _handleError(e);
    }
  }

  // Mark all notifications as read
  Future<int> markAllAsRead() async {
    try {
      final response = await _apiService.put(
        '${ApiConstants.notifications}/mark-all-read',
      );

      return response.data['count'] as int;
    } catch (e) {
      throw _handleError(e);
    }
  }

  // Delete notification
  Future<void> deleteNotification(int notificationId) async {
    try {
      await _apiService.delete(
        '${ApiConstants.notifications}/$notificationId',
      );
    } catch (e) {
      throw _handleError(e);
    }
  }

  String _handleError(dynamic error) {
    if (error is DioException) {
      if (error.response != null) {
        final data = error.response!.data;
        if (data is Map && data.containsKey('message')) {
          return data['message'];
        }
        return error.response!.statusMessage ?? 'An error occurred';
      }
      return 'Network error occurred';
    }
    return error.toString();
  }
}

File: frontend/src/app/services/notification.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, BehaviorSubject, interval } from 'rxjs';
import { tap, switchMap, startWith } from 'rxjs/operators';
import { API_CONFIG } from '../shared/api-config';

export interface Notification {
  id: number;
  type: string;
  title: string;
  message: string;
  entityType?: string;
  entityId?: number;
  read: boolean;
  createdAt: string;
}

@Injectable({
  providedIn: 'root'
})
export class NotificationService {
  private unreadCountSubject = new BehaviorSubject<number>(0);
  public unreadCount$ = this.unreadCountSubject.asObservable();

  constructor(private http: HttpClient) {
    // Poll for new notifications every 30 seconds
    interval(30000)
      .pipe(
        startWith(0),
        switchMap(() => this.getUnreadCount())
      )
      .subscribe();
  }

  getNotifications(page: number = 0, size: number = 20): Observable<any> {
    return this.http.get(`${API_CONFIG.NOTIFICATIONS}?page=${page}&size=${size}`);
  }

  getUnreadNotifications(): Observable<Notification[]> {
    return this.http.get<Notification[]>(`${API_CONFIG.NOTIFICATIONS}/unread`);
  }

  getUnreadCount(): Observable<{ count: number }> {
    return this.http.get<{ count: number }>(`${API_CONFIG.NOTIFICATIONS}/unread/count`).pipe(
      tap(response => this.unreadCountSubject.next(response.count))
    );
  }

  markAsRead(notificationId: number): Observable<any> {
    return this.http.put(`${API_CONFIG.NOTIFICATIONS}/${notificationId}/read`, {}).pipe(
      tap(() => {
        const currentCount = this.unreadCountSubject.value;
        this.unreadCountSubject.next(Math.max(0, currentCount - 1));
      })
    );
  }

  markAllAsRead(): Observable<any> {
    return this.http.put(`${API_CONFIG.NOTIFICATIONS}/mark-all-read`, {}).pipe(
      tap(() => this.unreadCountSubject.next(0))
    );
  }

  deleteNotification(notificationId: number): Observable<any> {
    return this.http.delete(`${API_CONFIG.NOTIFICATIONS}/${notificationId}`);
  }

  refreshUnreadCount(): void {
    this.getUnreadCount().subscribe();
  }
}
